resolve.go

```diff
@@ -0,0 +1,108 @@
+package matrixfederation
+
+import (
+	"net"
+	"strconv"
+	"strings"
+)
+
+// A HostResult is the result of looking up the IP addresses for a host.
+type HostResult struct {
+	CName string    // The canonical name for the host.
+	Addrs []string  // The IP addresses for the host.
+	Error *DNSError // If there was an error getting the IP addresses.
+}
+
+// A DNSResult is the result of looking up a matrix server in DNS.
+type DNSResult struct {
+	SRVCName   string                // The canonical name for the SRV record in DNS
+	SRVRecords []*net.SRV            // List of SRV record for the matrix server.
+	SRVError   *DNSError             // If there was an error getting the SRV records.
+	Hosts      map[string]HostResult // The results of looking up the SRV record targets.
+	Addrs      []string              // List of "<ip>:<port>" strings that the server is listening on.
+}
+
+// A DNSError describes an error that occured when trying to lookup a matrix server in DNS.
+type DNSError struct {
+	Error     string // The string decription of the error.
+	Temporary bool   // Whether the error is temporary.
+	Timeout   bool   // Whether the error was a timeout.
+}
+
+func dnsError(err error) (result *DNSError) {
+	if err == nil {
+		return
+	}
+	result = &DNSError{}
+	result.Error = err.Error()
+	if neterr, ok := err.(net.Error); ok {
+		result.Temporary = neterr.Temporary()
+		result.Timeout = neterr.Timeout()
+	}
+	return
+}
+
+// LookupServer looks up a matrix server in DNS.
+func LookupServer(serverName string) (result DNSResult, err error) {
```

Comment:
A 60 line function is waaaaay longer than I'd like if you're using named return args. Please, think of my sanity!